Added dynamic type and test cases

agamkrbit opened the pull request:
Made changes in the SDK to add dynamic type
- Now using github repository to fetch the test data
- Update `dto` to use `typedValue` instead of `value` and `typedOverrides` instead of `overrides`
- Added below function in the EppoClient to get typed values
  - getStringAssignment
  - getBooleanAssignment
  - getDoubleAssignment
  - getJsonAssignment
- Updated the test cases to check all the value types

aarsilv commented on Makefile:
```diff
@@ -26,13 +26,21 @@ help: Makefile
-	gsutil cp -r gs://sdk-test-data/assignment-v2 $(testDataDir)
+	mkdir -p $(tempDir)
+	cd ${tempDir} \
+	    && git clone -b ${branchName} --depth 1 --single-branch ${githubRepoLink} \
```
🔥

aarsilv commented on eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java:
```diff
@@ -154,18 +158,48 @@ public void testCachedAssignments() {
                         try {
-                            return client.getAssignment(subject.subjectKey, testCase.experiment, subject.subjectAttributes);
+                            switch (valueType) {
+                                case "numeric":
```
👆 Great callout Eric!

agamkrbit replied:
Added the enum!

aarsilv commented on eppo/src/main/java/cloud/eppo/android/dto/adapters/EppoValueAdapter.java:
```diff
@@ -34,24 +36,27 @@ public EppoValue deserialize(JsonElement json, Type typeOfT, JsonDeserialization
             try {
-                return EppoValue.valueOf(json.getAsInt());
-            } catch (Exception ignored) {}
+                return EppoValue.valueOf(json.getAsDouble());
```
I _think_, combined with the removal of `json.getAsLong()`, it's to support decimal numbers?

agamkrbit replied:
To support decimal. We are considering all the numeric type as double.

aarsilv reviewed (commented):
I don't know android, but I do know Java 😅 
Eric made lots of good suggestions here so I'll happily defer to him with nothing else to add.

leoromanovsky commented on Makefile:
```diff
@@ -26,13 +26,21 @@ help: Makefile
+githubRepoLink := https://github.com/Eppo-exp/sdk-test-data.git
 .PHONY: test-data
 test-data: 
 	rm -rf $(testDataDir)
```
`tempDir` is never created. I am going with in js repo:

```
	rm -rf $(testDataDir)
	rm -rf $(tempDir)
	mkdir -p $(testDataDir)
	mkdir -p $(tempDir)
```

agamkrbit replied:
When we create `tempDir ` is also creates `testDataDir ` because it is the sub-folder.

agamkrbit commented on eppo/src/main/java/cloud/eppo/android/EppoClient.java:
```diff
@@ -92,11 +96,11 @@ private Variation getAssignedVariation(String subjectKey, String experimentKey,
     }
 
-    public String getAssignment(String subjectKey, String flagKey, SubjectAttributes subjectAttributes) {
+    public EppoValue getAssignment(String subjectKey, String flagKey, SubjectAttributes subjectAttributes) {
```
Yes, make sense to make it private

agamkrbit commented on eppo/src/main/java/cloud/eppo/android/EppoClient.java:
```diff
@@ -128,17 +132,72 @@ public String getAssignment(String subjectKey, String flagKey, SubjectAttributes
-            assignmentLogger.logAssignment(assignment);
+         if (assignmentLogger != null) {
+         Assignment assignment = new Assignment(flagKey,
+         assignedVariation.getTypedValue().stringValue(), subjectKey,
```
My Bad, will fix it

agamkrbit commented on eppo/src/test/java/cloud/eppo/android/EppoValueDeserializerTest.java:
```diff
@@ -17,10 +17,10 @@ public class EppoValueDeserializerTest {
 
     @Test
-    public void testDeserializingInteger() throws Exception {
+    public void testDeserializingDouble() throws Exception {
```
Actually the new adapter treats all the numeric value as decimal. So, I have removed the older test. Do you think we should consider int and double differently?

leoromanovsky replied:
LaunchDarkly has separate methods for `getInt` and `getDouble` so we might need to add that in the future.

agamkrbit commented:
@leoromanovsky @petzel @aarsilv Made below changes in the code.
- Update the version to `1.0.0` from `0.3.0`
- Created enum `AssignmentValueType` for `valueType` field in `AssignmentTestCase` instead of keep it as string and also added deserialiser to convert json into enum.  Replace the switch cases with enum values instead of hard-coded string.
- keep the older `getAssignment` function as public and returning string value
- fixed the formatting issue

agamkrbit commented on eppo/src/main/java/cloud/eppo/android/EppoClient.java:
```diff
@@ -92,11 +96,11 @@ private Variation getAssignedVariation(String subjectKey, String experimentKey,
     }
 
-    public String getAssignment(String subjectKey, String flagKey) {
+    public EppoValue getAssignment(String subjectKey, String flagKey) {
```
Keep the value public and now returning string value.

agamkrbit commented on eppo/src/main/java/cloud/eppo/android/util/Converter.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Converter {
```
Moved the `Converter` to `androidTest` folder

agamkrbit commented:
> @agamkrbit thank you for adding the test cases from github this is going to be an awesome template for the future.
> 
> I am just noticing that unit tests are not run in github via actions. I have done android development in the past and realize that maybe there is no support for this in a cloud environment so for the purpose of this PR could you at the least show test passing from your local machine?
> 
> Personally I am a little stuck trying to get it to run on local:
> 
> ```
> 🔵 Uninstalling old version of test app(END)
> adb uninstall cloud.eppo.android.test
> adb: no devices/emulators found
> make: *** [test] Error 1
> ```
> 
> Ideally I would ask for your help:
> 
> * creating a github action to run `make test`
> * instructions for setting up a local environment
> 
> Thank you.

Sure @leoromanovsky . Make sense to add the local test screenshot. 
![Uploading Screenshot 2023-08-30 at 9.55.38 PM.png…]()

leoromanovsky commented on eppo/src/main/java/cloud/eppo/android/EppoClient.java:
```diff
@@ -128,17 +128,80 @@ public String getAssignment(String subjectKey, String flagKey, SubjectAttributes
+    }
+
+    public String getStringAssignment(String subjectKey, String flagKey, SubjectAttributes subjectAttributes) {
+        EppoValue value = this.getTypedAssignment(subjectKey, flagKey, subjectAttributes);
+        if (value == null) {
+            return null;
+        }
```
Mostly a style question in general - some of the other SDKs have a `Null` enum that the `EppoValue` can be - personally seems easier just to check against `null` like you do here. Any advantage that you see?

leoromanovsky replied:
Agreed thank you - I avoided it in the js changes https://github.com/Eppo-exp/js-client-sdk-common/pull/9
